ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AlterTableCreateMetaRefAnalyzer.java

```diff
@@ -0,0 +1,115 @@
+import org.apache.hadoop.hive.ql.parse.SemanticException;
+import org.apache.hadoop.hive.ql.session.SessionState;
+
+import static org.apache.hadoop.hive.ql.parse.AlterTableMetaRefSpec.AlterMetaRefOperationType.CREATE_BRANCH;
+import static org.apache.hadoop.hive.ql.parse.AlterTableMetaRefSpec.AlterMetaRefOperationType.CREATE_TAG;
+
+public abstract class AlterTableCreateMetaRefAnalyzer extends AbstractAlterTableAnalyzer {
+
+  private static AbstractAlterTableDesc alterTableDesc;
+
+  public AlterTableCreateMetaRefAnalyzer(QueryState queryState) throws SemanticException {
+    super(queryState);
+  }
+
+  @Override
+  protected void analyzeCommand(TableName tableName, Map<String, String> partitionSpec, ASTNode command)
+      throws SemanticException {
+    Table table = getTable(tableName);
+    DDLUtils.validateTableIsIceberg(table);
+    inputs.add(new ReadEntity(table));
+
+    String metaRefName = command.getChild(0).getText();
+    Long snapshotId = null;
+    Long asOfTime = null;
+    Long maxRefAgeMs = null;
+    Integer minSnapshotsToKeep = null;
+    Long maxSnapshotAgeMs = null;
+    AlterTableType alterTableType =
+        command.getType() == HiveParser.TOK_ALTERTABLE_CREATE_BRANCH ? AlterTableType.CREATE_BRANCH : AlterTableType.CREATE_TAG;
+    for (int i = 1; i < command.getChildCount(); i++) {
+      ASTNode childNode = (ASTNode) command.getChild(i);
+      switch (childNode.getToken().getType()) {
+      case HiveParser.TOK_AS_OF_VERSION:
+        snapshotId = Long.parseLong(childNode.getChild(0).getText());
+        break;
+      case HiveParser.TOK_AS_OF_TIME:
+        ZoneId timeZone = SessionState.get() == null ? new HiveConf().getLocalTimeZone() :
+            SessionState.get().getConf().getLocalTimeZone();
+        TimestampTZ ts = TimestampTZUtil.parse(stripQuotes(childNode.getChild(0).getText()), timeZone);
+        asOfTime = ts.toEpochMilli();
+        break;
+      case HiveParser.TOK_RETAIN:
+        String maxRefAge = childNode.getChild(0).getText();
+        String timeUnitOfBranchRetain = childNode.getChild(1).getText();
+        maxRefAgeMs =
+            TimeUnit.valueOf(timeUnitOfBranchRetain.toUpperCase(Locale.ENGLISH)).toMillis(Long.parseLong(maxRefAge));
+        break;
+      case HiveParser.TOK_WITH_SNAPSHOT_RETENTION:
+        minSnapshotsToKeep = Integer.valueOf(childNode.getChild(0).getText());
+        if (childNode.getChildren().size() > 1) {
+          String maxSnapshotAge = childNode.getChild(1).getText();
+          String timeUnitOfSnapshotsRetention = childNode.getChild(2).getText();
+          maxSnapshotAgeMs = TimeUnit.valueOf(timeUnitOfSnapshotsRetention.toUpperCase(Locale.ENGLISH))
+              .toMillis(Long.parseLong(maxSnapshotAge));
+        }
+        break;
+      default:
+        throw new SemanticException("Unrecognized token in ALTER " + alterTableType.getName() + " statement");
+      }
+    }
+
+    if (alterTableType == AlterTableType.CREATE_BRANCH) {
```

Comment:
why not keep it as abstract and create TAG and BRANCH implementations?

Reply:
Branch have more fields than Tag, we must handle Branch and tag respectively.

Reply:
sorry, I don't get it
````
class public abstract class AlterTableCreateMetaRefAnalyzer extends AbstractAlterTableAnalyzer {
    abstract AbstractAlterTableDesc  getAlterTableDesc(AlterTableTypeReq req)

    class AlterTableTypeReq{
        snapshotId, asOfTime, maxRefAgeMs, minSnapshotsToKeep, maxSnapshotAgeMs
    }
}

class AlterTableCreateBranchAnalyzer extends AlterTableCreateMetaRefAnalyzer {
    ...
    public AlterTableCreateBranchAnalyzer(QueryState queryState) throws SemanticException {
        super(queryState);
        alterTableType = HiveParser.TOK_ALTERTABLE_CREATE_BRANCH;
    }
 ....
    AbstractAlterTableDesc  getAlterTableDesc(AlterTableTypeReq req){
         AlterTableMetaRefSpec.CreateTagSpec createTagspec =
          new AlterTableMetaRefSpec.CreateTagSpec(req.getMetaRefName(), req.getSnapshotId(), req.getAsOfTime(), req.getMaxRefAgeMs());
      AlterTableMetaRefSpec<AlterTableMetaRefSpec.CreateTagSpec> alterTableTagSpec
          = new AlterTableMetaRefSpec(CREATE_TAG, createTagspec);
      return new AlterTableCreateTagDesc(tableName, alterTableTagSpec);
}
````

Reply:
Cool! Thanks for your guidance. Fixed!